### PR TITLE
Improve hardware abstraction layer

### PR DESCRIPTION
**Pull request by Mesteery**
This PR completly refactor the hardware abstraction layer. I think it's ready to be reviewed and I'm really sorry because it's a big PR and I didn't split the commits to ease the review. 
It compiled successfully on arduino cores (https://github.com/Mesteery/RadioLib/actions/runs/4680264029 and https://github.com/Mesteery/RadioLib/actions/runs/4681321278/jobs/8293761734 after stm32wlx fix)

- I was forced to remove `dir` default in methods like `setGdo0Action` because of the use of `this` (to access eg. `hal->GpioInterruptRising`)
- I'm not too happy with the RADIOLIB_NC check in the HAL because I don't think it's normal that the user has to check this kind of thing (in case he provides his own HAL). But this is the simplest and cleanest way, so it doesn't bother me more than that

*This is a breaking change.*

### CODE REVIEW

**Comment by jgromes** [comment]
Thanks, is this ready for review or still in progress?

**Comment by Mesteery** [comment]
It seems to be ready for review. I've tested it on my pi with real hardware, it also works!

**Comment by Mesteery** [comment]
By the way, could we use directly `uint8_t` instead of `byte` in the lib, so we cam remove this line for generic build: https://github.com/jgromes/RadioLib/blob/7faeb46ada637675e4c96677de370dd773435bc4/src/BuildOpt.h#L276

**Comment by jgromes** [comment]
As far as I can tell, there should be no usage of the `byte` type anywhere in the library, if there is then it's definitely a mistake. The only exception are the examples, which will remain Arduino-specific even after this is merged.

I'm a bit busy at the moment, but will try to do a proper review over the weekend.

File: src/BuildOpt.h [hunk 1]
```diff
@@ -7,21 +7,16 @@
   #define RADIOLIB_BUILD_ARDUINO
 #else
   // generic build
+  #include <stdio.h>
```
Comment (jgromes) [inline src/BuildOpt.h:10]: Why is this needed?

Comment (jgromes) [reply src/BuildOpt.h:10]: Similar with other includes (ctypes, math.h etc.) in other files.

File: src/BuildOpt.h [hunk 2]
```diff
@@ -105,833 +65,218 @@
...
+
+  #define DEC 10
+  #define HEX 16
+  #define OCT 8
+  #define BIN 2
+
+  #include <algorithm>
+  #include <stdint.h>
 
+  typedef uint8_t byte;
```
Comment (jgromes) [inline src/BuildOpt.h:276]: As per the conversation in the PR, this can likely be removed.

File: src/Hal.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#define _RADIOLIB_HAL_H
+
+class Hal {
+  public:
+    const uint8_t GpioModeInput;
+    const uint8_t GpioModeOutput;
+    const uint8_t GpioLevelLow;
+    const uint8_t GpioLevelHigh;
+    const uint8_t GpioInterruptRising;
+    const uint8_t GpioInterruptFalling;
```
Comment (jgromes) [inline src/Hal.h:13]: I don't think it's safe to use uint8_t as the generic "pin type", I would propose to change this to uint32_t as most of the platforms RadioLib runs on are 32-bit.

File: src/protocols/ExternalRadio/ExternalRadio.h [hunk 4]
```diff
@@ -3,12 +3,17 @@
 
 #include "../../TypeDef.h"
 #include "../../Module.h"
+#include "../../ArduinoHal.h"
```
Comment (jgromes) [inline src/protocols/ExternalRadio/ExternalRadio.h:6]: This should probably be included only for RADIOLIB_BUILD_ARDUINO.

File: src/Module.h [hunk 5]
```diff
@@ -1,8 +1,8 @@
+#include "TypeDef.h"
+#include "Hal.h"
```
Comment (jgromes) [inline src/Module.h:2]: Why is this moved before the guard ifdef?

File: src/Hal.cpp [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+      GpioLevelHigh(high),
+      GpioInterruptRising(rising),
+      GpioInterruptFalling(falling) {}
+
+void Hal::init(){};
+void Hal::term(){};
+void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
+void Hal::noTone(uint8_t pin){};
+void Hal::yield(){};
+uint8_t Hal::pinToInterrupt(uint8_t pin) { return pin; };
```
Comment (jgromes) [inline src/Hal.cpp:17]: What is the reason these methods are not pure virtual like the rest?

**Review by jgromes — CHANGES_REQUESTED** [review]
I left some comments, but it's nothing major. If everything can be implemented/answered, then I think this can be merged. Thanks!

**Comment by jgromes** [comment]
Also, I noticed there's now a couple warnings in the CI (building for Arduino Uno), related to the (pure) virtual comment I left.

```
Building AFSK_Imperial_March.ino ... 
/home/runner/work/RadioLib/RadioLib/src/Hal.cpp: In member function 'virtual void Hal::tone(uint8_t, unsigned int, long unsigned int)':
/home/runner/work/RadioLib/RadioLib/src/Hal.cpp:14:24: warning: unused parameter 'pin' [-Wunused-parameter]
 void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
                        ^~~
/home/runner/work/RadioLib/RadioLib/src/Hal.cpp:14:42: warning: unused parameter 'frequency' [-Wunused-parameter]
 void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
                                          ^~~~~~~~~
/home/runner/work/RadioLib/RadioLib/src/Hal.cpp:14:67: warning: unused parameter 'duration' [-Wunused-parameter]
 void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
                                                                   ^~~~~~~~
/home/runner/work/RadioLib/RadioLib/src/Hal.cpp: In member function 'virtual void Hal::noTone(uint8_t)':
/home/runner/work/RadioLib/RadioLib/src/Hal.cpp:15:26: warning: unused parameter 'pin' [-Wunused-parameter]
 void Hal::noTone(uint8_t pin){};
                          ^~~
Sketch uses 20276 bytes (62%) of program storage space. Maximum is 32256 bytes.
Global variables use 990 bytes (48%) of dynamic memory, leaving 1058 bytes for local variables. Maximum is 2048 bytes.
```

File: src/BuildOpt.h (hunk 1 above)
Comment (Mesteery) [reply src/BuildOpt.h:10]: There indeed useless for Arduino build since they are already included, but on generic build they are required to use some functions like `sprinf`, `max`, `min`, `fabs`, etc.
These includes did not cause any problems with arduino, so I left them unconditioned at the top of the file using functions from them.

File: src/BuildOpt.h (hunk 2 above)
Comment (Mesteery) [reply src/BuildOpt.h:276]: Yes I forgot to remove it, since there is yet some files using `byte` in the lib.

File: src/protocols/ExternalRadio/ExternalRadio.h (hunk 4 above)
Comment (Mesteery) [reply src/protocols/ExternalRadio/ExternalRadio.h:6]: ArduinoHal.h would be empty if RADIOLIB_BUILD_ARDUINO is empty, so is it important to re-check?

File: src/Module.h (hunk 5 above)
Comment (Mesteery) [reply src/Module.h:2]: I don't remember much about it, but maybe it's because some define were not found, or maybe it's completely by mistake. I remember at least one case like that but I don't know where. I'll try again.

File: src/Hal.cpp (hunk 6 above)
Comment (Mesteery) [reply src/Hal.cpp:17]: They are not pure because on some platform they doesn't exist, and will simply be defined as nop. We can simply make them optional and implement them as not for platform that doesn't support them.
init and term are optional since they are `Hal`-specific and can or not be useful.

**Comment by Mesteery** [comment]
> Also, I noticed there's now a couple warnings in the CI (building for Arduino Uno), related to the (pure) virtual comment I left.
> 
> ```
> Building AFSK_Imperial_March.ino ... 
> /home/runner/work/RadioLib/RadioLib/src/Hal.cpp: In member function 'virtual void Hal::tone(uint8_t, unsigned int, long unsigned int)':
> /home/runner/work/RadioLib/RadioLib/src/Hal.cpp:14:24: warning: unused parameter 'pin' [-Wunused-parameter]
>  void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
>                         ^~~
> /home/runner/work/RadioLib/RadioLib/src/Hal.cpp:14:42: warning: unused parameter 'frequency' [-Wunused-parameter]
>  void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
>                                           ^~~~~~~~~
> /home/runner/work/RadioLib/RadioLib/src/Hal.cpp:14:67: warning: unused parameter 'duration' [-Wunused-parameter]
>  void Hal::tone(uint8_t pin, unsigned int frequency, unsigned long duration){};
>                                                                    ^~~~~~~~
> /home/runner/work/RadioLib/RadioLib/src/Hal.cpp: In member function 'virtual void Hal::noTone(uint8_t)':
> /home/runner/work/RadioLib/RadioLib/src/Hal.cpp:15:26: warning: unused parameter 'pin' [-Wunused-parameter]
>  void Hal::noTone(uint8_t pin){};
>                           ^~~
> Sketch uses 20276 bytes (62%) of program storage space. Maximum is 32256 bytes.
> Global variables use 990 bytes (48%) of dynamic memory, leaving 1058 bytes for local variables. Maximum is 2048 bytes.
> ```

I left them because I find writing `(void)var;` not pretty like in some files to remove warning. But finally I will just do like this, it's better than useless warning :+1:

File: src/Hal.h (hunk 3 above)
Comment (Mesteery) [reply src/Hal.h:13]: I don't understand , why wouldn't it be safe?

File: src/BuildOpt.h [hunk 7]
```diff
@@ -105,833 +65,218 @@
...
+
+  #if !defined(RADIOLIB_ARDUINOHAL_PIN_MODE_CAST)
+  #define RADIOLIB_ARDUINOHAL_PIN_MODE_CAST
+  #endif
+  #if !defined(RADIOLIB_ARDUINOHAL_PIN_STATUS_CAST)
+  #define RADIOLIB_ARDUINOHAL_PIN_STATUS_CAST
+  #endif
+  #if !defined(RADIOLIB_ARDUINOHAL_INTERRUPT_MODE_CAST)
+  #define RADIOLIB_ARDUINOHAL_INTERRUPT_MODE_CAST
+  #endif
```
Comment (Mesteery) [inline src/BuildOpt.h:253]: (Not important again) I wanted to know if it's not too "ugly" the repetition of `ifndef .. define .. endif`. It seems to be quite syntactically heavy but I haven't found an alternative.

File: src/BuildOpt.h (hunk 2 above)
Comment (jgromes) [reply src/BuildOpt.h:276]: > there is yet some files using byte in the lib.

Where exactly is this?

File: src/protocols/ExternalRadio/ExternalRadio.h (hunk 4 above)
Comment (jgromes) [reply src/protocols/ExternalRadio/ExternalRadio.h:6]: At the moment, yes, it would be empty, but I would rather avoid this breaking in the future. Also IIRC there's other places where ArduinoHal.h is included which are guarded, so let's make it consistent. It's a compile-time check anyway.

**Comment by jgromes** [comment]
> I left them because I find writing (void)var; not pretty like in some files to remove warning

I would say that in these cases it's justified - actually, the only times when (void)var; is used is when it's a parameters that is needed for some common API, which is also teh case here.

File: src/BuildOpt.h (hunk 7 above)
Comment (jgromes) [reply src/BuildOpt.h:253]: I would prefer it like this, if the user decides to define the macro in some other way (for example by calling gcc with -DRADIOLIB_ARDUINOHAL_INTERRUPT_MODE_CAST=something).

File: src/BuildOpt.h (hunk 1 above)
Comment (jgromes) [reply src/BuildOpt.h:10]: Thanks for the examplanation, in that case this can remain.

File: src/Hal.h (hunk 3 above)
Comment (jgromes) [reply src/Hal.h:13]: > I don't think there is a platform with more than 255 pins

But the value can contain more information than just a pin number. In fact most MCUs address GPIO pins using port and pin type. IIRC some STM32 Arduino cores use the number to save both GPIO port as well as GPIO pin, which may add up to more than 8 bits.

File: src/Hal.cpp (hunk 6 above)
Comment (jgromes) [reply src/Hal.cpp:17]: OK, thanks for the clarification, this can remain as-is.

File: src/BuildOpt.h (hunk 2 above)
Comment (Mesteery) [reply src/BuildOpt.h:276]: https://github.com/jgromes/RadioLib/blob/88549baf8326ee9f679e09e04f7b3c1a0e8d7a2f/src/protocols/Morse/Morse.h#L127 via gh code search.

Comment (jgromes) [reply src/BuildOpt.h:276]: Ah, I see - that should be `uint8_t`, feel free to change it (or I can do after merging this).

File: src/BuildOpt.h (hunk 7 above)
Comment (Mesteery) [reply src/BuildOpt.h:253]: Yes sure, I was rather searching a different syntax or way to do that.

Comment (jgromes) [reply src/BuildOpt.h:253]: As far as I know there's nothing else, I've seen this syntax quite a lot.

**Review by jgromes — APPROVED** [review]
All looking good now, thanks!

**Comment by jgromes** [comment]
@Mesteery I was about to merge this, but then I realized it could also be useful to have an example of how to use the new HAL. Since you already have yours, maybe you could add it to e.g. `examples/Non_Arduino/...`? We would have to modify the CI file so that it doesn't try to build it as Arduino example, but that is pretty straight-forward (it has a filter).

Or we can merge this as-is and open another PR for the example, either way is fine if you're willing to share it of course.

**Comment by jgromes** [comment]
Perfect, then I can try it right away using RPi and a Waveshare LoRaWAN hat!

**Comment by Mesteery** [comment]
Do you have any news, I'm curious to know if it worked!

**Comment by jgromes** [comment]
@Mesteery unfrotunately not yet, recently I haven't had enough time to put into this project :( Hoping to get this done over the weekend again, sorry it's taking longer.

**Comment by jgromes** [comment]
@Mesteery I fainlly got around to merge this - thank you very much for the great contribution!

In the following days, I will do a bit of cleanup and then release this as 6.0.0.
